crates/bevy_picking/src/mesh_picking/ray_cast/intersections.rs

```diff
@@ -0,0 +1,405 @@
+                distance: mesh_transform
+                    .transform_vector3(mesh_space_ray.direction * hit.distance)
+                    .length(),
+                triangle: hit.triangle.map(|tri| {
+                    [
+                        mesh_transform.transform_point3a(tri[0]),
+                        mesh_transform.transform_point3a(tri[1]),
+                        mesh_transform.transform_point3a(tri[2]),
+                    ]
+                }),
+                triangle_index,
+            });
+            closest_hit_distance = hit.distance;
+        }
+    }
+
+    closest_hit
+}
+
+#[inline(always)]
+fn triangle_intersection(
+    tri_vertices: [Vec3A; 3],
+    tri_normals: Option<[Vec3A; 3]>,
+    max_distance: f32,
+    ray: &Ray3d,
+    backface_culling: Backfaces,
+) -> Option<RayMeshHit> {
+    let hit = ray_triangle_intersection(ray, &tri_vertices, backface_culling)?;
+
+    if hit.distance < 0.0 || hit.distance > max_distance {
+        return None;
+    };
+
+    let point = ray.get_point(hit.distance);
+    let u = hit.barycentric_coords.0;
+    let v = hit.barycentric_coords.1;
+    let w = 1.0 - u - v;
+    let barycentric = Vec3::new(u, v, w);
+
+    let normal = if let Some(normals) = tri_normals {
+        normals[1] * u + normals[2] * v + normals[0] * w
+    } else {
+        (tri_vertices[1] - tri_vertices[0])
+            .cross(tri_vertices[2] - tri_vertices[0])
+            .normalize()
+    };
+
+    Some(RayMeshHit {
+        point,
+        normal: normal.into(),
+        barycentric_coords: barycentric,
+        distance: hit.distance,
+        triangle: Some(tri_vertices),
+        triangle_index: None,
+    })
+}
+
+/// Takes a ray and triangle and computes the intersection.
+#[inline(always)]
+fn ray_triangle_intersection(
+    ray: &Ray3d,
+    triangle: &[Vec3A; 3],
+    backface_culling: Backfaces,
+) -> Option<RayTriangleHit> {
+    // Source: https://www.scratchapixel.com/lessons/3d-basic-rendering/ray-tracing-rendering-a-triangle/moller-trumbore-ray-triangle-intersection
+    let vector_v0_to_v1: Vec3A = triangle[1] - triangle[0];
+    let vector_v0_to_v2: Vec3A = triangle[2] - triangle[0];
+    let p_vec: Vec3A = (Vec3A::from(*ray.direction)).cross(vector_v0_to_v2);
+    let determinant: f32 = vector_v0_to_v1.dot(p_vec);
+
+    match backface_culling {
+        Backfaces::Cull => {
+            // if the determinant is negative the triangle is back facing
+            // if the determinant is close to 0, the ray misses the triangle
+            // This test checks both cases
+            if determinant < f32::EPSILON {
+                return None;
+            }
+        }
+        Backfaces::Include => {
+            // ray and triangle are parallel if det is close to 0
+            if determinant.abs() < f32::EPSILON {
+                return None;
+            }
+        }
+    }
+
+    let determinant_inverse = 1.0 / determinant;
```

Comment:
is it possible `determinant` is ever 0?

Reply:
I believe it could happen if the ray is coplanar with the triangle (passing through the surface of the triangle and multiple points on a line), as that would produce a `p_vec` value which is orthogonal to `vector_v0_to_v2`, making their dot product zero.

Reply:
I think it is possible, but the `!(0.0..=1.0).contains(&u)` check should return early in that case, so it shouldn't return any infinite or NaN values